File: lib/text_decoder.dart
// lib/text_decoder.dart (New File)
import 'gpu_tensor.dart';
import 'layer_norm.dart';
import 'nn.dart';
import 'aft_text_decoder_block.dart'; // Your new file

class TextDecoder extends Module {
  final int vocabSize;
  final int maxSeqLen;
  final int embedSize;

  final Tensor wte; // Token Embeddings
  final Tensor wpe; // Positional Embeddings
  final List<TextDecoderBlock> blocks;
  final LayerNorm finalLayerNorm;
  final Layer lmHead; // Language Model Head (embedSize -> vocabSize)

  TextDecoder({
    required this.vocabSize,
    required this.maxSeqLen,
    required this.embedSize,
    int numLayers = 4,
    int numHeads = 4,
  }) : wte = Tensor.random([vocabSize, embedSize], scale: 0.02),
       wpe = Tensor.random([maxSeqLen, embedSize], scale: 0.02),
       blocks = List.generate(
         numLayers,
         (i) => TextDecoderBlock(embedSize, numHeads, maxSeqLen),
       ),
       finalLayerNorm = LayerNorm(embedSize),
       lmHead = Layer(
         embedSize,
         vocabSize,
         useGelu: false,
       ); // Project to vocabulary

  Tensor forward(
    List<int> textTokens,
    Tensor encoderOutput,
    List<Tensor> tracker,
  ) {
    final int T = textTokens.length;
    if (T > maxSeqLen) {
      throw ArgumentError("Text sequence length exceeds maxSeqLen");
    }

    // 1. Embed text tokens and add positional information
    final tokenEmbeds = Tensor.embeddings(textTokens, wte, wpe);
    tracker.add(tokenEmbeds);

    // 2. Pass through decoder blocks
    Tensor x = tokenEmbeds;
    for (final block in blocks) {
      x = block.forward(x, encoderOutput, tracker);
    }

    // 3. Final LayerNorm and Language Model Head
    final normalized = finalLayerNorm.forward(x, tracker);
    return lmHead.forward(
      normalized,
      tracker,
    ); // Logits for each vocabulary token
  }

  @override
  List<Tensor> parameters() => [
    wte,
    wpe,
    ...blocks.expand((block) => block.parameters()),
    ...finalLayerNorm.parameters(),
    ...lmHead.parameters(),
  ];
}

File: python/dart_cuda/core/transformers/modalities/text_decoder.py
"""Text decoder — mirrors `modalities/text_decoder.dart`."""

from __future__ import annotations

from typing import List, Sequence

from ...layers.layer_norm import LayerNorm
from ...layers.nn import Layer, Module
from ...tensor.gpu_tensor import Tensor
from ..aft.text_decoder_block import TextDecoderBlock


class TextDecoder(Module):
    def __init__(
        self,
        *,
        vocabSize: int,
        maxSeqLen: int,
        embedSize: int,
        numLayers: int = 4,
        numHeads: int = 4,
    ):
        self.vocab_size = vocabSize
        self.max_seq_len = maxSeqLen
        self.embed_size = embedSize
        self.wte = Tensor.random([vocabSize, embedSize], scale=0.02)
        self.wpe = Tensor.random([maxSeqLen, embedSize], scale=0.02)
        self.blocks = [
            TextDecoderBlock(embedSize, numHeads, maxSeqLen)
            for _ in range(numLayers)
        ]
        self.final_layer_norm = LayerNorm(embedSize)
        self.lm_head = Layer(embedSize, vocabSize, useGelu=False)

    def forward(self, text_tokens: Sequence[int], encoder_output: Tensor, tracker: List[Tensor]) -> Tensor:
        T = len(text_tokens)
        if T > self.max_seq_len:
            raise ValueError("Text sequence length exceeds maxSeqLen")
        x = Tensor.embeddings(text_tokens, self.wte, self.wpe)
        tracker.append(x)
        for block in self.blocks:
            x = block.forward(x, encoder_output, tracker)
        normalized = self.final_layer_norm.forward(x, tracker)
        return self.lm_head.forward(normalized, tracker)

    def parameters(self) -> List[Tensor]:
        ps: List[Tensor] = [self.wte, self.wpe]
        for b in self.blocks:
            ps.extend(b.parameters())
        ps.extend(self.final_layer_norm.parameters())
        ps.extend(self.lm_head.parameters())
        return ps
